clients/venice-client/src/test/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClientTest.java

```diff
@@ -0,0 +1,313 @@
+      assertFalse(metrics.get("." + STORE_NAME + metricPrefix + "retry_request_success_key_count.Rate").value() > 0);
+      assertFalse(batchGetRequestContext.numberOfKeysCompletedInRetryRequest.get() > 0);
+    } else {
+      assertFalse(metrics.get("." + STORE_NAME + "--error_retry_request.OccurrenceRate").value() > 0);
+      assertFalse(getRequestContext.errorRetryRequestTriggered);
+      assertFalse(metrics.get("." + STORE_NAME + "--long_tail_retry_request.OccurrenceRate").value() > 0);
+      assertFalse(getRequestContext.longTailRetryRequestTriggered);
+      assertFalse(metrics.get("." + STORE_NAME + "--retry_request_win.OccurrenceRate").value() > 0);
+      assertFalse(getRequestContext.retryWin);
+    }
+  }
+
+  private byte[] serializeBatchGetResponse(Set<String> Keys) {
+    List<MultiGetResponseRecordV1> routerRequestValues = new ArrayList<>(Keys.size());
+    AtomicInteger count = new AtomicInteger();
+    Keys.stream().forEach(key -> {
+      MultiGetResponseRecordV1 routerRequestValue = new MultiGetResponseRecordV1();
+      byte[] valueBytes =
+          dispatchingAvroGenericStoreClient.getKeySerializer().serialize(BATCH_GET_VALUE_RESPONSE.get(key));
+      ByteBuffer valueByteBuffer = ByteBuffer.wrap(valueBytes);
+      routerRequestValue.setValue(valueByteBuffer);
+      routerRequestValue.keyIndex = count.getAndIncrement();
+      routerRequestValues.add(routerRequestValue);
+    });
+    return dispatchingAvroGenericStoreClient.getMultiGetSerializer().serializeObjects(routerRequestValues);
+  }
+
+  @Test
+  public void testGet() throws ExecutionException, InterruptedException, IOException {
+    try {
+      setUpClient();
+      getRequestContext = new GetRequestContext();
+      String value = statsAvroGenericStoreClient.get(getRequestContext, "test_key").get().toString();
+      assertEquals(value, SINGLE_GET_VALUE_RESPONSE);
+      metrics = getStats(clientConfig);
+      assertTrue(metrics.get("." + STORE_NAME + "--healthy_request.OccurrenceRate").value() > 0);
+      assertTrue(metrics.get("." + STORE_NAME + "--healthy_request_latency.Avg").value() > 0);
+      assertFalse(metrics.get("." + STORE_NAME + "--unhealthy_request.OccurrenceRate").value() > 0);
+      assertFalse(metrics.get("." + STORE_NAME + "--unhealthy_request_latency.Avg").value() > 0);
+      assertFalse(metrics.get("." + STORE_NAME + "--no_available_replica_request_count.OccurrenceRate").value() > 0);
+      assertFalse(getRequestContext.noAvailableReplica);
+      assertEquals(metrics.get("." + STORE_NAME + "--request_key_count.Max").value(), 1.0);
+      assertEquals(metrics.get("." + STORE_NAME + "--success_request_key_count.Max").value(), 1.0);
+      assertEquals(getRequestContext.successRequestKeyCount.get(), 1);
+
+      validateRetryMetrics();
+    } finally {
+      tearDown();
+    }
+  }
+
+  @Test
+  public void testGetWithExceptionFromTransportLayer() throws IOException {
```

Comment:
Is it possible to also write an unit test to reproduce the issue we saw where rejected requests (sending to a host that's not available) is not reported as unhealthy requests?

Reply:
I made a quick attempt at it and looks like I need to spend some time figuring out whats happening there. Will probably add as a separate test and its respective fix later.